packages/shift-next/src/pages/product.js

```diff
@@ -30,15 +32,55 @@ class ProductPage extends Component {
     this.addToBag = this.addToBag.bind(this)
   }
 
-  static async getInitialProps ({ reduxStore, req, query }) {
-    const { id } = query
+  static async getInitialProps ({ reduxStore, req, res, query: { id } }) {
+    const product = await ProductPage.fetchProduct(id, reduxStore.dispatch)
     const isServer = !!req
+    const { sellable } = product
+    let preview
 
-    if (isServer) {
-      await reduxStore.dispatch(readProduct(id))
+    if (req && req.query) {
+      preview = req.query.preview
+    } else {
+      preview = false
+    }
+
+    // Unpublished products should return a 404 page to the public however we should
+    // be able to preview unpublished products using a ?preview=true query string
+    if (sellable === false && preview === false) {
+      if (isServer) {
+        // server-side
+        res.redirect('/error')
+      } else {
+        // client-side
+        Router.push('/error')
+      }
+    }
+
+    return { id: id, product }
+  }
+
+  static async fetchProduct (id, dispatch) {
+    try {
+      const request = ProductPage.productRequest(id)
+      const response = await new ApiClient().read(request.endpoint, request.query, dispatch)
```

Comment:
I am not a massive fan of passing redux.dispatch around everywhere - it's not really the pattern that redux encourages.  I don't have a better suggestion for now tho so 🤷‍♂️

Reply:
I basically followed the pattern we are using in `static-page.js`. Happy to refactor all pages if/when we nail a better pattern down.